### ISSUE TITLE
Altscreen: rendering artifacts when resizing window

### ISSUE THREAD

**Issue by purpleclay** [charmbracelet/bubbletea#573]
Raising this issue based on the discussion with @muesli #544

It appears that with some terminals if you use `lipgloss` to fill available space with a background colour it automatically pads with whitespace. This causes a rendering glitch when you resize the terminal window due to word wrapping. A terminal such as Alacritty doesn't exhibit this behaviour. From testing, it exhibits this behaviour with iTerm2 and the native Mac Terminal.app.

https://user-images.githubusercontent.com/106762954/196023116-87c03990-5678-4c9e-95f7-46c6dbfcb074.mov

A potential solution is to fill the available height of the terminal with empty lines to prevent this wrapping behaviour.

Example App:

```go
package main

import (
	"log"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

var (
	PrimaryColour = lipgloss.Color("#3a1577")

	Banner = lipgloss.NewStyle().Background(PrimaryColour).Height(2)
)

type model struct {
	width int
	text  string
}

func (m model) Init() tea.Cmd {
	return nil
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	cmds := make([]tea.Cmd, 0)

	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width = msg.Width
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c":
			return m, tea.Quit
		}
	}

	return m, tea.Batch(cmds...)
}

func (m model) View() string {
	var b strings.Builder

	b.WriteString(Banner.Copy().Width(m.width).Render(m.text))
	return b.String()
}

func main() {
	m := model{
		text: "hello!",
	}

	if err := tea.NewProgram(m, tea.WithAltScreen()).Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Comment by kristiandrucker** [comment]
I'm having the same issue and am just using viewort without lipgloss. Could this be an underlying issue in the renderer?

**Comment by meowgorithm** [comment]
This is indeed an improvement that can be made to the renderer: thanks for flagging the issue. In the meantime, you can workaround the issue by manually clearing the screen on resize.

```diff
diff --git a/main.go b/main.go
index 426c6cb..a777967 100644
--- a/main.go
+++ b/main.go
@@ -29,6 +29,7 @@ func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 	switch msg := msg.(type) {
 	case tea.WindowSizeMsg:
 		m.width = msg.Width
+		return m, ClearScreen
 	case tea.KeyMsg:
 		switch msg.String() {
 		case "ctrl+c":
```

**Comment by kristiandrucker** [comment]
@meowgorithm thanks for the tip. I've had a couple issues in my code and was fixing it. The fix for now, although I don't know if it really fixes it is to set the viewPort (for me) to the actual required height and cap it at the max allowed height (terminal - any lines outside of the viewPort)

**Comment by meowgorithm** [comment]
@kristiandrucker Noted, thank you. We'll keep this issue updated with progress.

**Comment by StarpTech** [comment]
Hi, any updates? The workaround works, but seeing a fix here would be great.

**Comment by StarpTech** [comment]
This https://github.com/charmbracelet/bubbletea/issues/573#issuecomment-1342855822 doesn't work for my issue https://github.com/charmbracelet/bubbletea/issues/717

**Comment by meowgorithm** [comment]
@StarpTech, noted thank you. In terms of the renderer, the fix should apply to both altscreen and non-altscreen cases.

As an additional note, everyone, this is pretty common thing you’ll encounter with many TUIs and, in some cases, it’s unfixable. If you resize vim enough you’ll encounter artifacts as well.

**Comment by StarpTech** [comment]
Hi @meowgorithm thank you for the insights. That sounds reasonable but I have never seen the UI re-rendered multiple times on scroll so that the console buffer is growing all the time (non alt-mode). Check my video in my issue.

**Comment by meowgorithm** [comment]
@StarpTech Yep, saw it—thanks for posting. While we have some hypotheses on how to improve the situation for cases like this, inline solutions will be harder to solve for because Bubble Tea does not have control of the entire window for repaints the way it does with the altscreen open. That said, inline rendering is of first-class concern to Bubble Tea and we do want to give that case the same attention as altscreen. Given that, I'll reopen your issue so we can track progress accordingly.

Also keep in mind that we're a small team and, while we attempt to move as quickly as possible with issues we're still a fairly small team.

Nice work on the wundergraph TUI, btw!

**Comment by StarpTech** [comment]
Any updates?
